src/linear_algebra/projection.lean

```diff
@@ -330,3 +332,69 @@ def is_compl_equiv_proj :
   (p.is_compl_equiv_proj.symm f : submodule R E) = (f : E →ₗ[R] p).ker := rfl
 
 end submodule
+
+namespace linear_map
+
+open submodule
+
+/--
+A linear endomorphism of a module `E` is a projection onto a submodule `p` if it sends every element
+of `E` to `p` and fixes every element of `p`.
+-/
+structure is_proj (f : E →ₗ[R] E) : Prop :=
```

Comment:
(I mention this because we have another PR defining projections, and that version doesn't mention a submodule

Reply:
What is F here?

Reply:
Some arbitrary module

Reply:
I'm not sure how we can define projections like this. For the condition `f x = x` to make sense we need the codomain to be the same as the domain, or a submodule of it (which is how they are defined the rest of the file, with `p` as the codomain).

Reply:
Right, I think I was misremembering the other PR. It's #12173, but I was struggling to find it on mobile. Can you take a quick look at that to tell if you think anything can be shared between the PRs?

Reply:
The definition `is_projection f` in #12173 is equivalent to `∃ p, is_proj p f` when `f` is of type `E →ₗ[R] E`, but for the applications I have in mind I find it useful to have `p` explicitly in the definition. Other than that definition there doesn't seem to be any overlap, and since the definition `is_projection` is for any monoid I think it's fine to keep them separated.

Reply:
Though I think I'll add a proof in my PR that `∃ p, is_proj p f` is equivalent to `f^2=f`.

Reply:
And in theother direction, `f^2=f` implies `is_proj f.range f`, right?

Reply:
Yes